optuna/storages/_journal/storage.py

```diff
@@ -0,0 +1,608 @@
+            "trial_id": trial_id,
+            "state": state,
+            "values": values,
+        }
+
+        if state == TrialState.RUNNING:
+            log["datetime_start"] = datetime.datetime.now().isoformat()
+        elif state.is_finished():
+            log["datetime_complete"] = datetime.datetime.now().isoformat()
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_STATE_VALUES, log)
+            self._sync_with_backend()
+
+            if (
+                state == TrialState.RUNNING
+                and trial_id not in self._replay_result._trial_ids_owned_by_this_process
+            ):
+                return False
+            else:
+                return True
+
+    def set_trial_intermediate_value(
+        self, trial_id: int, step: int, intermediate_value: float
+    ) -> None:
+        log: Dict[str, Any] = {
+            "trial_id": trial_id,
+            "step": step,
+            "intermediate_value": intermediate_value,
+        }
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_INTERMEDIATE_VALUE, log)
+            self._sync_with_backend()
+
+    def set_trial_user_attr(self, trial_id: int, key: str, value: Any) -> None:
+        log: Dict[str, Any] = {
+            "trial_id": trial_id,
+            "user_attr": {key: value},
+        }
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_USER_ATTR, log)
+            self._sync_with_backend()
+
+    def set_trial_system_attr(self, trial_id: int, key: str, value: Any) -> None:
+        log: Dict[str, Any] = {
+            "trial_id": trial_id,
+            "system_attr": {key: value},
+        }
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_SYSTEM_ATTR, log)
+            self._sync_with_backend()
+
+    def get_trial(self, trial_id: int) -> FrozenTrial:
+        with self._thread_lock:
+            self._sync_with_backend()
+            return self._replay_result.get_trial(trial_id)
+
+    def get_all_trials(
+        self,
+        study_id: int,
+        deepcopy: bool = True,
+        states: Optional[Container[TrialState]] = None,
+    ) -> List[FrozenTrial]:
+        with self._thread_lock:
+            self._sync_with_backend()
+            frozen_trials = self._replay_result.get_all_trials(study_id, states)
+            if deepcopy:
+                return copy.deepcopy(frozen_trials)
+            return frozen_trials
+
+
+class JournalStorageReplayResult:
+    def __init__(self, pid: str) -> None:
+        self.log_number_read = 0
+        self._pid = pid
+        self._studies: Dict[int, FrozenStudy] = {}
+        self._trials: Dict[int, FrozenTrial] = {}
+
+        self._study_id_to_trial_ids: Dict[int, List[int]] = {}
+        self._trial_id_to_study_id: Dict[int, int] = {}
+        self._next_study_id: int = 0
+        self._trial_ids_owned_by_this_process: List[int] = []
```

Comment:
We should rename `_trial_ids_owned_by_this_process` to `_trial_ids_created_by_this_process` because this list holds trials which will be just created by this process. The reason I emphasized this name is that, when we enqueue a trial, it is created by this process but not necessarily owned by it.

Reply:
Addressed in [b67e0b5](https://github.com/optuna/optuna/pull/3854/commits/b67e0b53d4aa33ab0e0bc64a96942732deb1c0c6).